packages/panzer/disc-fe/src/evaluators/Panzer_Integrator_CurlBasisDotVector_impl.hpp

```diff
@@ -40,424 +40,776 @@
-
-    // multiply agains all the fields in the next one
-    for (typename std::vector<PHX::MDField<const ScalarT,Cell,IP> >::iterator field = field_multipliers.begin();
-         field != field_multipliers.end(); ++field) {
-      fillScratch.field = *field;
-
-      Kokkos::parallel_for(Kokkos::RangePolicy<PHX::Device,typename FillScratch::FieldMultipliers>(0,workset.num_cells), fillScratch);
-    }
-
-    // Build integrate values functor (hard code spatial dimensions)
-    IntegrateValuesVector<ScalarT,3> intValues;
-    intValues.scratch     = scratch_vector;
-    intValues.residual    = residual;
-    intValues.weighted_curl_basis = bv.weighted_curl_basis_vector;
-    
-    Kokkos::parallel_for(workset.num_cells, intValues);
-  }
-  else {
-    typedef FillScratchScalar<ScalarT> FillScratch;
-    FillScratch fillScratch;
-    fillScratch.scratch     = scratch_scalar;
-    fillScratch.multiplier  = multiplier;
-    fillScratch.vectorField = flux_scalar;
-
-    // initialize in first pass (basically a scaled copy)
-    Kokkos::parallel_for(Kokkos::RangePolicy<PHX::Device,typename FillScratch::Initialize>(0,workset.num_cells), fillScratch);
-
-    // multiply agains all the fields in the next one
-    for (typename std::vector<PHX::MDField<const ScalarT,Cell,IP> >::iterator field = field_multipliers.begin();
-         field != field_multipliers.end(); ++field) {
-      fillScratch.field = *field;
-
-      Kokkos::parallel_for(Kokkos::RangePolicy<PHX::Device,typename FillScratch::FieldMultipliers>(0,workset.num_cells), fillScratch);
-    }
-
-    // Build integrate values functor
-    IntegrateValuesScalar<ScalarT> intValues;
-    intValues.scratch     = scratch_scalar;
-    intValues.residual    = residual;
-    intValues.weighted_curl_basis = bv.weighted_curl_basis_scalar;
-
-    Kokkos::parallel_for(workset.num_cells, intValues);
-  }
-
-#if 0
-  residual.deep_copy(ScalarT(0.0));
-
-  for (index_t cell = 0; cell < workset.num_cells; ++cell)
+    /**
+     *  \brief Multiply the integrand by the scalar multiplier (\f$ M \f$) and
+     *         any field multipliers (\f$ a(x) \f$, \f$ b(x) \f$, etc.) out in
+     *         front of the integral.
+     *
+     *  This must happen before the integration itself is performed by
+     *  `Integrate2D`.
+     *
+     *  \note This class is used when integrating a two-dimensional problem. 
+     *        Three-dimensional problems use the `PreMultiply3D` class.
+     */
+    template<typename ScalarT>
+    class PreMultiply2D
+    {
+      public:
+
+        /**
+         *  \brief This empty struct allows us to template the `operator()()`
+         *         routine on it, such that one call can pre-multiply the
+         *         integrand by the scalar multiplier and another call can
+         *         pre-multiply it by a field multiplier.
+         */
+        struct ScalarMultiplier
```

Comment:
We usually call these dispatch "tags" in kokkos Maybe add the "Tag" qualifier to these (ScalarMultiplier --> ScalarMultiplierTag). This is how I have been doing it. You could also then  trim down the verbosity in the comment to just say it is the kokkos dispatch tag. Whatever is your preference.

Reply:
Good thought.